normalizer_lib.py

```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env python
+"""Library of language-agnostic FST rewrite rules to normalize text."""
+
+import unicodedata
+from pynini import *
+from config import *
+
+LANGUAGE = af
+
+GRAPHEMES = LANGUAGE.GRAPHEMES
+
+INITIAL_PUNCTUATION = LANGUAGE.INITIAL_PUNCTUATION
+FINAL_PUNCTUATION = LANGUAGE.FINAL_PUNCTUATION
+OTHER_PUNCTUATION = union(r"\[", r"\]")
+PUNCTUATION = union(INITIAL_PUNCTUATION, FINAL_PUNCTUATION, OTHER_PUNCTUATION)
+NUMBERS = union("0", "1", "2", "3", "4", "5", "6", "7", "8", "9")
+SPACE = acceptor(" ")
+
+SIGMA_STAR = union(*("[{}]".format(i) for i in range(1, 256))
+                   ).optimize().closure()
+
+
+# Step 1: Remove all extra whitespace between words
+
+REMOVE_EXTRA_WHITESPACE = transducer(SPACE, "")
+
+DO_REMOVE_EXTRA_WHITESPACE = cdrewrite(
+    REMOVE_EXTRA_WHITESPACE,
+    "",
+    SPACE,
+    SIGMA_STAR)
+
+
+# Step 2: Language-specific formatting fixes
+
+DO_LANGUAGE_SPECIFIC_PREPROCESSING = LANGUAGE.LANGUAGE_SPECIFIC_PREPROCESSING
+
+
+# Step 3: Apply NFC unicode normalize
+# really slow when done this way,
+# super fast when I just call unicodedata.normalize
+# on the string at the beginning
+# may remove later
+#def unicode_normalize(string: str) -> Fst:
+#    "Normalizes unicode to NFC normalization."
+#    nfc_normalized = unicodedata.normalize("NFC", string)
+#    normalizer = transducer(string, nfc_normalized).optimize()
+#    return cdrewrite(normalizer, "", "", SIGMA_STAR).optimize()
+
+
+# Step 4: Discard examples not associated with a pronunciation
+# e.g. "how are you today товарищ?" -> "how are you today?"
+
+SENTENCE = (INITIAL_PUNCTUATION.ques + (GRAPHEMES.plus | NUMBERS.plus) + FINAL_PUNCTUATION.ques +
+            (SPACE.plus + GRAPHEMES.plus).star + FINAL_PUNCTUATION.ques)
+
+
+def pass_only_valid(string: str) -> str:
+    "Allows only valid tokens of the language to pass through."
+    valid_token = GRAPHEMES.plus
```

Comment:
This will currently reject any email addresses, times, websites, words that are number/letter combos like "mp3", etc.  This works for this inital pull request, so no need to make changes before we merge this.  Perhaps a good next step for a follow up merge could be to define what we want to accept by way of these not-a-word tokens.  You could define acceptors for these like:

email_address = GRAPHEMES.plus + "@" + GRAPHEMES.plus + closure("." + GRAPHEMES.plus, 1, 2)

This would accept the following:
crewe@google.com
crewe@google.co.uk
etc.

Then you put them all together by saying something like:
def verbalizable(self):
  return email_address @ web_address

Then you include "self.verbalizable" in your token acceptor for passing valid tokens.

Reply:
I'll look into adding this! Also it turns out that `valid_token = GRAPHEMES.plus` isn't actually used in this function. The comparison actually uses 
```SENTENCE = (INITIAL_PUNCTUATION.ques + (GRAPHEMES.plus | NUMBERS.plus) + FINAL_PUNCTUATION.ques + (SPACE.plus + GRAPHEMES.plus).star + FINAL_PUNCTUATION.ques)```
which is a little bit fancier than just a sequence of graphemes, but I will for sure try to add the other valid tokens like email addresses etc.

Reply:
I added a new class at the beginning of `normalizer_lib` to handle verbalizable tokens. It currently includes `EMAIL_ADDRESS`, `WEB_ADDRESS`, `TIME`, and  `FANCY_NUMBER` (which is for numbers with decimal/grouping separators like `.` and `,`). It also allows combinations of letters and numbers in tokens, so we now allow things like `mp3`.

We can follow-up on what other not-a-word tokens we want to include in a later PR.